### PR TITLE
arch/arm/src/armv7-m: Add support of nested interrupts

### PR DESCRIPTION
**Pull request by drashti304**
This patch adds the support of nested interrupts for armv7-m architecture.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [5bac1ad2364622696df10e47516b995e026e8a52] -

**Comment by seinfra** [comment]
Target : [5bac1ad2364622696df10e47516b995e026e8a52] -

File: os/arch/arm/src/armv7-m/up_lazyexception.S [hunk 1]
```diff
@@ -245,8 +245,12 @@ exception_common:
 	 * here prohibits nested interrupts without some additional logic!
 	 */
 
-	setintstack	r2, r3
-
+	ldr		r5, =g_nestlevel
+	ldr		r6, [r5]
```
Comment (sijupunnnoose) [inline os/arch/arm/src/armv7-m/up_lazyexception.S]: Can you leave a comment over here, "If g_nestlevel is zero then behave as normal, switching from the user to the interrupt stack; if g_nestlevel is greater than zero, then do not switch stacks. In this latter case, we are already using the interrupt stack."

**Comment by seinfra** [comment]
Target : [b4e12fc8dd3d2aa94ebd3d85702a031812d4403c] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [b4e12fc8dd3d2aa94ebd3d85702a031812d4403c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ee21034519ffc61a13770a88dc7b3b55947b7e49] -

**Comment by seinfra** [comment]
Target : [ee21034519ffc61a13770a88dc7b3b55947b7e49] -

**Comment by seinfra** [comment]
Target : [7ebab95af093643f99b44edcea3cb3a40e1f1373] -

**Comment by seinfra** [comment]
Target : [7ebab95af093643f99b44edcea3cb3a40e1f1373] -

File: os/arch/arm/src/armv7-m/up_lazyexception.S (hunk 1 above)
Comment (drashti304) [reply os/arch/arm/src/armv7-m/up_lazyexception.S]: Done

File: os/arch/arm/src/armv7-m/up_lazyexception.S [hunk 2]
```diff
@@ -210,7 +210,7 @@ exception_common:
 	 * and call up_doirq to handle the interrupt
 	 */
 
-	cpsid	i						/* Disable further interrupts */
+	/*cpsid   i */				/* Disable further interrupts */
```
Comment (sangam-swami) [inline os/arch/arm/src/armv7-m/up_lazyexception.S]: can you use proper format or combine these two in one comment

File: os/arch/arm/src/armv7-m/up_lazyexception.S [hunk 3]
```diff
@@ -371,7 +375,7 @@ exception_common:
 #ifdef CONFIG_ARMV7M_USEBASEPRI
 	msr		basepri, r3				/* Restore interrupts priority masking */
 #ifndef CONFIG_ARCH_HIPRI_INTERRUPT
-	cpsie	i						/* Re-enable interrupts */
+	/*cpsie   i*/						/* Re-enable interrupts */
```
Comment (sangam-swami) [inline os/arch/arm/src/armv7-m/up_lazyexception.S]: same as above

**Comment by seinfra** [comment]
Target : [4a0284040a39ac4afbd4cb41c2a25f31b34d2720] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4a0284040a39ac4afbd4cb41c2a25f31b34d2720] - Code Rule Check OK.

File: os/arch/arm/src/armv7-m/up_lazyexception.S (hunk 2 above)
Comment (drashti304) [reply os/arch/arm/src/armv7-m/up_lazyexception.S]: modified the code to make it macro protected

File: os/arch/arm/src/armv7-m/up_lazyexception.S (hunk 3 above)
Comment (drashti304) [reply os/arch/arm/src/armv7-m/up_lazyexception.S]: modified the code to make it macro protected

File: os/arch/arm/src/armv7-m/up_doirq.c [hunk 4]
```diff
@@ -101,22 +125,43 @@ uint32_t *up_doirq(int irq, uint32_t *regs)
...
 	/* Acknowledge the interrupt */
 
 	up_ack_irq(irq);
+#endif
 
 	/* Deliver the IRQ */
 
 	irq_dispatch(irq, regs);
 
+
```
Comment (amandeepc) [inline os/arch/arm/src/armv7-m/up_doirq.c]: Please remove this extra line

File: os/arch/Kconfig [hunk 5]
```diff
@@ -571,6 +572,20 @@ config ARCH_INTERRUPTSTACK
...
+	default n
+
+if ARCH_INTERRUPTSTACK > 7
+config ARCH_NESTED_INTERRUPT
+	bool "Nested interrupt support"
+	depends on ARCH_HAVE_NESTED_INTERRUPT
+	default n
+	---help---
+		Enables the nested interrupt support. If defined, higher priority interrupt
+		can pre-empte the lower priority interrupt.
```
Comment (sangam-swami) [inline os/arch/Kconfig]: pre-empt

File: os/arch/arm/src/armv7-m/up_doirq.c [hunk 6]
```diff
@@ -88,10 +90,32 @@
...
+	/* Current regs non-zero indicates that we are processing an interrupt;
+	 * regs holds the state of the interrupted logic; current_regs holds the
+	 * state of the interrupted user task.  current_regs should, therefor,
+	 * only be modified for outermost interrupt handler (when g_nestlevel == 0)
+	 */
+
+	flags = irqsave();
+
+	if (g_nestlevel == 0)
+	{
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-m/up_doirq.c]: coding rule

File: os/arch/arm/src/armv7-m/up_doirq.c [hunk 7]
```diff
@@ -101,22 +125,43 @@ uint32_t *up_doirq(int irq, uint32_t *regs)
...
+	 * switch occurred during interrupt processing.  Context switching should
+	 * only be performed when the outermost interrupt handler returns.
+	 */
+
+	flags = irqsave();
+
+	g_nestlevel--;
+
+	if (g_nestlevel == 0)
+	{
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-m/up_doirq.c]: same as above

File: os/arch/Kconfig [hunk 8]
```diff
@@ -13,6 +13,7 @@ config ARCH_ARM
 	select ARCH_HAVE_VFORK
 	select ARCH_HAVE_STACKCHECK
 	select ARCH_HAVE_CUSTOMOPT
+	select ARCH_HAVE_NESTED_INTERRUPT
```
Comment (sunghan-chang) [inline os/arch/Kconfig]: You only modify armv7-m. Is it ok for armv7-r with this?

File: os/arch/arm/src/armv7-m/up_doirq.c (hunk 4 above)
Comment (drashti304) [reply os/arch/arm/src/armv7-m/up_doirq.c]: Removed

**Comment by seinfra** [comment]
Target : [86f482920a004e110b53310fd7072d0698906bb4] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [86f482920a004e110b53310fd7072d0698906bb4] - Code Rule Check (C++) OK.

File: os/arch/Kconfig (hunk 8 above)
Comment (drashti304) [reply os/arch/Kconfig]: Currently I have implemented for armv7-m only. We will check feasibility for armv7-r later.

File: os/arch/arm/src/armv7-m/up_doirq.c (hunk 6 above)
Comment (drashti304) [reply os/arch/arm/src/armv7-m/up_doirq.c]: Done

File: os/arch/arm/src/armv7-m/up_doirq.c (hunk 7 above)
Comment (drashti304) [reply os/arch/arm/src/armv7-m/up_doirq.c]: Done

File: os/arch/Kconfig (hunk 8 above)
Comment (sunghan-chang) [reply os/arch/Kconfig]: 1. So, it does not work with armv7-r. But you added selection under ARM, not armv7-m. Is it ok to use armv7-r now?
2. Please let me know the ETA of armv7-r.

**Comment by sunghan-chang** [comment]
@drashti304 Please keep a file permission with 644. You changed two files.

File: os/arch/Kconfig (hunk 8 above)
Comment (drashti304) [reply os/arch/Kconfig]: Hi @sunghan-chang, 

1) I checked armv7-r is working fine as "ARCH_NESTED_INTERRUPT" macro is disable by default.
2) I will be able to complete the changes of armv7-r most probable by 5th june

**Comment by drashti304** [comment]
> @drashti304 Please keep a file permission with 644. You changed two files.

I have reverted the file permission to 644

**Comment by seinfra** [comment]
Target : [ca0dbbcde80cb8bbdf7d8cb117020fa69427bae8] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ca0dbbcde80cb8bbdf7d8cb117020fa69427bae8] - Code Rule Check (C++) OK.

**Pull request merged**